research/checksum-verification-proposal.md

```diff
@@ -0,0 +1,364 @@
+        dep, _ := kclPkg.Dependencies.Deps.Get(name)
+        
+        // Generate new checksum
+        newChecksum, err := HashDir(dep.LocalFullPath)
+        if err != nil {
+            return fmt.Errorf("failed to generate new checksum for %s: %w", dep.Name, err)
+        }
+
+        // Update checksum in Dependency struct
+        dep.Checksum = newChecksum
+
+        // Update checksum in kcl.mod.lock
+        err = writeChecksumToLockFile(dep, newChecksum)
+        if err != nil {
+            return fmt.Errorf("failed to update checksum in lock file for %s: %w", dep.Name, err)
+        }
+    }
+}
+```
+
+4. Verify checksums when vendoring dependencies
+
+https://github.com/kcl-lang/kpm/blob/2b1e0b91d8488da804158728063308ca271a3350/pkg/client/client.go#L953
+
+```go
+func (c *KpmClient) VendorDeps(kclPkg *pkg.KclPkg) error {
+    // Existing code...
+
+    for _, name := range kclPkg.Dependencies.Deps.Keys() {
+        dep, _ := kclPkg.Dependencies.Deps.Get(name)
+        if !c.noSumCheck {
+            err := verifyChecksum(dep)
+            if err != nil {
+                return fmt.Errorf("checksum verification failed for %s during vendoring: %w", dep.Name, err)
+            }
+        }
+        // Existing vendoring logic...
+    }
+}
+```
+
+5. Verify checksums before compilation
+
+https://github.com/kcl-lang/kpm/blob/2b1e0b91d8488da804158728063308ca271a3350/pkg/client/client.go#L492
+
+```go
+func (c *KpmClient) Compile(kclPkg *pkg.KclPkg, opts *opt.CompileOptions) error {
+    // Verify checksums before compilation
+    if !c.noSumCheck {
+        for _, name := range kclPkg.Dependencies.Deps.Keys() {
+            dep, _ := kclPkg.Dependencies.Deps.Get(name)
+            err := verifyChecksum(dep)
+            if err != nil {
+                return fmt.Errorf("checksum verification failed for %s before compilation: %w", dep.Name, err)
+            }
+        }
+    }
+}
+```
+
+These changes will integrate checksum verification throughout the KPM workflow, ensuring the integrity of dependencies during addition, resolution, updating, vendoring, and compilation processes. The `FLAG_NO_SUM_CHECK` is respected in all cases, allowing for backward compatibility and gradual rollout of the checksum verification feature.
+
+## Compatibility solution
+
+To maintain compatibility with existing KPM dependencies while introducing checksum verification:
+
+- Implement a gradual rollout strategy using the existing FLAG_NO_SUM_CHECK.
+- Generate log warnings for packages without checksums initially.
+- Make checksum verification optional during a transition period.
+- After the transition period, make checksum verification mandatory for all packages.
+
+## Checker Module
+
+To improve modularity and testability, we'll introduce a separate Checker module for validating KCL dependencies. This module will handle name validation, version checking, and checksum verification.
+
+### Checker Module Design
```

Comment:
See https://github.com/kcl-lang/kpm/pull/431#discussion_r1717817452 for more details.